perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -2673,4 +2725,116 @@ private void processMemberAfterRemovingLastSponsor(PerunSession sess, Member spo
 			throw new InternalErrorException(e);
 		}
 	}
+
+	/**
+	 * Creates a new user from given data and sponsors him in the given vo.
+	 *
+	 * @param sess session
+	 * @param vo vo, where the new user will be sponsored
+	 * @param namespace namespace used to define an external system where
+	 *                  the user will have a new login generated (currently, only 'mu' namespace is supported)
+	 * @param data values used to create the new user.
+	 *             Required values are - firstname, lastname, urn:perun:user:attribute-def:def:preferredMail
+	 *             Optional values are - urn:perun:user:attribute-def:def:note
+	 * @param sponsor user, who will be set as a sponsor to the newly created user
+	 * @param validityTo validity of the sponsorship. If null, the sponsorship will not be automatically canceled.
+	 * @param asyncValidation switch for easier testing
+	 * @return result of the procedure
+	 */
+	private Map<String, String> createSingleSponsoredMemberFromCSV(PerunSession sess, Vo vo, String namespace,
+	                                                               Map<String, String> data, User sponsor,
+	                                                               LocalDate validityTo, boolean asyncValidation) {
+		for (String requiredField : SPONSORED_MEMBER_REQUIRED_FIELDS) {
+			if (!data.containsKey(requiredField)) {
+				log.error("Invalid data passed, missing required value: {}", requiredField);
+				throw new InternalErrorException("Invalid data passed, missing required value: " + requiredField);
+			}
+		}
+
+		Set<String> additionalValues = new HashSet<>(data.keySet());
+		additionalValues.removeAll(SPONSORED_MEMBER_REQUIRED_FIELDS);
+
+		for (String valueName : additionalValues) {
+			if (!SPONSORED_MEMBER_ADDITIONAL_FIELDS.contains(valueName)) {
+				log.error("Not allowed additional value passed, value: {}", valueName);
+				throw new InternalErrorException("Not allowed additional value passed, value: " + valueName);
+			}
+		}
+
+		Map<String, String> mapName = new HashMap<>();
+		mapName.put("firstName", data.get("firstname"));
+		mapName.put("lastName", data.get("lastname"));
+
+		String email = data.get(A_U_PREF_MAIL);
+
+		PasswordManagerModule module = getPerunBl().getUsersManagerBl().getPasswordManagerModule(sess, namespace);
+		String password = module.generateRandomPassword(sess, null);
+
+		// create sponsored member
+		User user;
```

Comment:
Is there any reason to declare user here and not later?

Reply:
No, there is not. I have moved it.